### PR TITLE
Ignore curl error on failed proxy authentication.

### LINKED ISSUE #12015: ProxyAuth test fails on *Nix outerloop runs
PR #11985 added a new test which is failing on *Nix (part of OuterLoop):

`System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test.UseCallback_HaveNoCredsAndUseAuthenticatedCustomProxyAndPostToSecureServer_ProxyAuthenticationRequiredStatusCode`

```
System.Net.Http.CurlException : Failure when receiving data from the peer
```

### PR DESCRIPTION
**Pull request by Priya91**
fixes #12015 

Curl behaves differently when the proxy server returns 407 Authentication failure for the cases of connecting to a https server vs http server. Curl throws error 56 in case of https server and 0 in case of http server. Hence parsing for this difference and making .NET not throw CURL exception in case of https to match the http case and the Windows behavior.

cc @stephentoub @davidsh @CIPop @geoffkizer @karelz @steveharter

### CODE REVIEW

**Comment by Priya91** [comment]
@dotnet-bot Test outerloop centos7.1 debug
@dotnet-bot test outerloop ubuntu16.04 release

**Comment by Priya91** [comment]
@dotnet-bot test outerloop osx debug

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 1]
```diff
@@ -849,7 +849,14 @@ private void FinishRequest(StrongToWeakReference<EasyRequest> easyWrapper, CURLc
                     // ignore it and treat such failures as successes, to match the Windows behavior.
                     if (messageResult != CURLcode.CURLE_UNSUPPORTED_PROTOCOL)
                     {
-                        ThrowIfCURLEError(messageResult);
+                        // libcurl will return CURLE_RECV_ERROR (56) if proxy authentication failed when connecting to a https server,
+                        // whereas it returns CURLE_OK for a http server proxy authentication failure. We ignore this curl behavior error,
+                        // and let the user rely on response message status code to match the Windows behavior.
+                        if (messageResult != CURLcode.CURLE_RECV_ERROR ||
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Is this going to "cover up" any errors where we'd actually want to throw an exception for this error code?

File: src/Common/src/Interop/Unix/System.Net.Http.Native/Interop.CURLcode.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@ internal enum CURLcode
             CURLE_OPERATION_TIMEDOUT = 28,
             CURLE_ABORTED_BY_CALLBACK = 42,
             CURLE_UNKNOWN_OPTION = 48,
+            CURLE_RECV_ERROR = 56,
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Net.Http.Native/Interop.CURLcode.cs]: Please add corresponding entries in the native files:
https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Net.Http.Native/pal_easy.h#L66
https://github.com/dotnet/corefx/blob/master/src/Native/Unix/System.Net.Http.Native/pal_easy.cpp#L49

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 1 above)
Comment (Priya91) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: the exception will be thrown for this curl error code for all cases except when there's a 407 http status code. The recv error is thrown when there's a network or remote server problem, in which case there would be no response from server. So i dont think this particular condition will mask any other error.

**Comment by Priya91** [comment]
The outerloop centos test failure is unrelated to this change.

File: src/Common/src/Interop/Unix/System.Net.Http.Native/Interop.CURLcode.cs (hunk 2 above)
Comment (Priya91) [inline src/Common/src/Interop/Unix/System.Net.Http.Native/Interop.CURLcode.cs]: Added.

**Pull request merged**